File: frontend/src/lib/hook-run-display.test.ts
import { beforeAll, describe, expect, it } from 'vitest'

import { i18n } from '../i18n/runtime'
import {
  formatHookRunHandlerLabel,
  formatHookOutputEntryText,
  formatHookRunFeedbackEntries,
  formatHookRunDecision,
  formatHookRunEventName,
  formatHookRunMessage,
  formatHookRunReason,
  formatHookRunStatus,
  formatHookRunTriggerMethodLabel,
  formatHookRunToolLabel,
  formatSessionStartSource,
} from './hook-run-display'

describe('hook-run-display', () => {
  beforeAll(() => {
    i18n.loadAndActivate({ locale: 'en', messages: {} })
  })

  it('formats dedicated thread entry hook event names with readable labels', () => {
    expect(formatHookRunEventName('TurnStart')).toBe('Turn Start')
    expect(formatHookRunEventName('TurnSteer')).toBe('Turn Steer')
    expect(formatHookRunEventName('TurnInterrupt')).toBe('Turn Interrupt')
    expect(formatHookRunEventName('ReviewStart')).toBe('Review Start')
  })

  it('formats common governance reasons with readable labels', () => {
    expect(formatHookRunReason('turn_start_audited')).toBe('Turn start audited')
    expect(formatHookRunReason('steer_no_active_turn')).toBe('Steer requested without an active turn')
    expect(formatHookRunReason('validation_command_failed')).toBe('Validation command failed')
    expect(formatHookRunReason('mcp_elicitation_request_audited')).toBe(
      'MCP elicitation request audited',
    )
    expect(
      formatHookRunReason('protected_governance_file_mutation_observed_after_mcp_tool_call'),
    ).toBe('Protected governance file mutation observed after MCP tool call')
    expect(formatHookRunReason('config_mcp_server_reload_audited')).toBe(
      'MCP server reload audited',
    )
  })

  it('formats hook run status and decision labels', () => {
    expect(formatHookRunStatus('inProgress')).toBe('In progress')
    expect(formatHookRunDecision('continueTurn')).toBe('Continue Turn')
    expect(formatSessionStartSource('resume')).toBe('Resume')
  })

  it('formats hook run tool labels with readable names', () => {
    expect(formatHookRunToolLabel('command/exec')).toBe('Command Execution')
    expect(formatHookRunToolLabel(undefined, 'commandExecution')).toBe('Command Execution')
    expect(formatHookRunToolLabel('filesystem/write_file')).toBe('Filesystem / Write File')
    expect(formatHookRunToolLabel('mcp/filesystem/exec_command')).toBe(
      'MCP / Filesystem / Exec Command',
    )
  })

  it('formats hook run handler labels with readable names', () => {
    expect(formatHookRunHandlerLabel('builtin.posttooluse.failed-validation-rescue')).toBe(
      'Failed Validation Rescue',
    )
    expect(formatHookRunHandlerLabel('builtin.pretooluse.block-dangerous-command')).toBe(
      'Dangerous Command Guard',
    )
    expect(formatHookRunHandlerLabel('builtin.httpmutation.audit-workspace-mutation')).toBe(
      'Workspace Mutation Audit',
    )
  })

  it('formats hook run trigger labels with readable names', () => {
    expect(formatHookRunTriggerMethodLabel('item/completed')).toBe('Item Completed')
    expect(formatHookRunTriggerMethodLabel('turn/start')).toBe('Turn Start')
    expect(formatHookRunTriggerMethodLabel('mcpServer/elicitation/request')).toBe(
      'MCP Elicitation Request',
    )
    expect(formatHookRunTriggerMethodLabel('item/commandExecution/requestApproval')).toBe(
      'Command Execution Approval Request',
    )
    expect(formatHookRunTriggerMethodLabel('fs/write')).toBe('Write File')
  })

  it('formats structured hook output entry text with readable labels', () => {
    expect(formatHookOutputEntryText('sourcePath=.codex/hooks.json')).toBe(
      'Source Path: .codex/hooks.json',
    )
    expect(formatHookOutputEntryText('matched policy: broad-recursive-delete')).toBe(
      'Matched Policy: broad-recursive-delete',
    )
    expect(
      formatHookRunFeedbackEntries([
        { kind: 'feedback', text: 'command=go test ./...; status=failed; exitCode=1' },
        { kind: 'context', text: 'matched path: docs/governance.md' },
      ]),
    ).toBe(
      'Command: go test ./...; status=failed; exitCode=1 | Matched Path: docs/governance.md',
    )
  })

  it('builds readable hook run summary messages', () => {
    expect(
      formatHookRunMessage({
        eventName: 'PostToolUse',
        handlerKey: 'builtin.posttooluse.failed-validation-rescue',
        triggerMethod: 'item/completed',
        status: 'completed',
        decision: 'continueTurn',
        toolName: 'command/exec',
        reason: 'validation_command_failed',
      }),
    ).toBe(
      'Event: Post-Tool Use\nHandler: Failed Validation Rescue\nStatus: Completed\nDecision: Continue Turn\nTrigger: Item Completed\nTool: Command Execution\nReason: Validation command failed',
    )
  })

  it('falls back to humanized labels for unmapped values', () => {
    expect(formatHookRunEventName('CustomHookEvent')).toBe('Custom Hook Event')
    expect(formatHookRunReason('custom_reason_code')).toBe('Custom Reason Code')
  })

  it('includes session start source in readable hook run summary messages', () => {
    expect(
      formatHookRunMessage({
        eventName: 'SessionStart',
        handlerKey: 'builtin.sessionstart.inject-project-context',
        status: 'completed',
        decision: 'continue',
        triggerMethod: 'turn/start',
        reason: 'project_context_injected',
        sessionStartSource: 'clear',
      }),
    ).toBe(
      'Event: Session Start\nHandler: Project Context Injection\nStatus: Completed\nDecision: Continue\nTrigger: Turn Start\nSession Start Source: Clear\nReason: Project context injected',
    )
  })
})

File: backend/internal/store/hook_run_display_test.go
package store

import "testing"

func TestFormatHookOutputEntryTextUsesReadableLabelsForStructuredValues(t *testing.T) {
	t.Parallel()

	if got := FormatHookOutputEntryText("sourcePath=.codex/hooks.json"); got != "Source Path: .codex/hooks.json" {
		t.Fatalf("expected readable source path label, got %q", got)
	}
	if got := FormatHookOutputEntryText("matched policy: broad-recursive-delete"); got != "Matched Policy: broad-recursive-delete" {
		t.Fatalf("expected readable matched policy label, got %q", got)
	}
}

func TestFormatHookRunFeedbackEntriesUsesReadableLabels(t *testing.T) {
	t.Parallel()

	got := FormatHookRunFeedbackEntries([]HookOutputEntry{
		{Kind: "feedback", Text: "command=go test ./...; status=failed; exitCode=1"},
		{Kind: "context", Text: "matched path: docs/governance.md"},
	}, 2)

	want := "Command: go test ./...; status=failed; exitCode=1 | Matched Path: docs/governance.md"
	if got != want {
		t.Fatalf("expected formatted feedback %q, got %q", want, got)
	}
}

func TestFormatHookRunReasonUsesDedicatedLabelsForMcpAndAuditReasons(t *testing.T) {
	t.Parallel()

	cases := map[string]string{
		"mcp_elicitation_request_audited":                                 "MCP elicitation request audited",
		"protected_governance_file_mutation_observed_after_mcp_tool_call": "Protected governance file mutation observed after MCP tool call",
		"dynamic_tool_call_request_audited":                               "Dynamic tool call request audited",
		"config_mcp_server_reload_audited":                                "MCP server reload audited",
	}

	for input, want := range cases {
		if got := FormatHookRunReason(input); got != want {
			t.Fatalf("expected reason %q to format as %q, got %q", input, want, got)
		}
	}
}

func TestFormatHookRunToolLabelUsesReadableNames(t *testing.T) {
	t.Parallel()

	cases := []struct {
		toolName string
		toolKind string
		want     string
	}{
		{toolName: "command/exec", want: "Command Execution"},
		{toolKind: "commandExecution", want: "Command Execution"},
		{toolName: "filesystem/write_file", want: "Filesystem / Write File"},
		{toolName: "mcp/filesystem/exec_command", want: "MCP / Filesystem / Exec Command"},
	}

	for _, tc := range cases {
		if got := FormatHookRunToolLabel(tc.toolName, tc.toolKind); got != tc.want {
			t.Fatalf("expected tool label %q/%q to format as %q, got %q", tc.toolName, tc.toolKind, tc.want, got)
		}
	}
}

func TestFormatHookRunHandlerLabelUsesReadableNames(t *testing.T) {
	t.Parallel()

	cases := map[string]string{
		"builtin.posttooluse.failed-validation-rescue":  "Failed Validation Rescue",
		"builtin.pretooluse.block-dangerous-command":    "Dangerous Command Guard",
		"builtin.httpmutation.audit-workspace-mutation": "Workspace Mutation Audit",
		"builtin.sessionstart.inject-project-context":   "Project Context Injection",
		"builtin.stop.require-successful-verification":  "Successful Verification Requirement",
	}

	for input, want := range cases {
		if got := FormatHookRunHandlerLabel(input); got != want {
			t.Fatalf("expected handler label %q to format as %q, got %q", input, want, got)
		}
	}
}

func TestFormatHookRunTriggerMethodLabelUsesReadableNames(t *testing.T) {
	t.Parallel()

	cases := map[string]string{
		"item/completed":                        "Item Completed",
		"turn/start":                            "Turn Start",
		"mcpServer/elicitation/request":         "MCP Elicitation Request",
		"item/commandExecution/requestApproval": "Command Execution Approval Request",
		"fs/write":                              "Write File",
		"config/mcp-server/reload":              "MCP Server Reload",
		"hook/follow-up":                        "Hook Follow-up",
	}

	for input, want := range cases {
		if got := FormatHookRunTriggerMethodLabel(input); got != want {
			t.Fatalf("expected trigger method %q to format as %q, got %q", input, want, got)
		}
	}
}
